### PR TITLE
Let-Binding

### PR DESCRIPTION
**Pull request by bechang**
This is a new pull-request to merge the let-binding feature for jsy. This pull request replaces #17.

### CODE REVIEW

File: src/test/scala/edu/colorado/plv/cuanto/jsy/binding/BindingParserSpec.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+         ),
+    // statements
+    "x; y" -> Binary(Seq, Var("x"), Var("y"))
+  )
+
+  val flexibles = Table(
+    "concrete" -> "abstract",
+    "{ let x = y y }"
+      -> Bind(Var("x"), Var("y"), Var("y"))
+  )
```
Comment (kyleheadley) [inline src/test/scala/edu/colorado/plv/cuanto/jsy/binding/BindingParserSpec.scala:46]: If I read the code correctly, it looks like 
"{ let x = y y y }" -> Binary(Seq, Bind(Var("x"), Var("y"), Var("y")), Var("y"),
or just: "{x y}" -> Binary(Seq, Var("x"), Var("y"))
Is this correct/intended/acceptable?

File: src/main/scala/edu/colorado/plv/cuanto/jsy/binding/Parser.scala [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+/** Parse variables and bindings.
+  *
+  * @author Bor-Yuh Evan Chang
+  */
+trait ParserLike extends OpParserLike with JsyParserLike {
+  import ParserLike._
+
+  override def stmt: Parser[Expr] = {
+    rep(concreteStmt) ^^ { stmts =>
+      stmts.foldRight(None: Option[Expr]) {
```
Comment (kyleheadley) [inline src/main/scala/edu/colorado/plv/cuanto/jsy/binding/Parser.scala:17]: I'm a little worried that accepting concatenations of arbitrary expressions will make this statement parser difficult or confusing for modular use.

**Review by kyleheadley — COMMENTED** [review]
As the parsers become more complex, do we need to create a doc the goes over the composition expectations? These are certainly very composable, but order of composition seems to be gaining requirements. Perhaps this is all standard, but we don't have a higher-order structure guiding these modular units.

**Comment by kyleheadley** [comment]
My previous comments are probably things that can be saved for later. I think the code as is is acceptable for getting us ready for more interpreter work.

**Review by bennostein — APPROVED** [review]
Looks good to me.

File: src/main/scala/edu/colorado/plv/cuanto/jsy/binding/Parser.scala (hunk 2 above)
Comment (bechang) [reply src/main/scala/edu/colorado/plv/cuanto/jsy/binding/Parser.scala:17]: As I think I said in #17, I am a bit torn about this, as the alternative would be to treat \n as a token. The statement-expression layering helps some, but it does allow some of these statement blocks that seem a bit odd.

**Comment by bechang** [comment]
@kyleheadley I agree with you that there needs to be documentation on how these parser modules should go together. I added some more references from the `arithmetic` one, as I think that would be a natural starting point. Certainly, I would your suggestions in where and how to document (or even going in and adding some of your documentation).

**Comment by bechang** [comment]
@kyleheadley I'm going to go ahead and do the merge here. If there's something that you want to make note for a future revision, please create an issue.

File: src/main/scala/edu/colorado/plv/cuanto/jsy/binding/Parser.scala (hunk 2 above)
Comment (bennostein) [reply src/main/scala/edu/colorado/plv/cuanto/jsy/binding/Parser.scala:17]: I think that this statement-expression bifurcation (which allows concatenation of arbitrary expressions) is the right solution here - getting into the business of dealing with newline characters and whitespace is messier, in my opinion.

**Pull request merged**